TESTING.asciidoc

```diff
@@ -76,7 +76,177 @@ In order to start a node with a trial license execute the following command:
+It is typically extended directly by unit tests.
+
+The Gradle target to run unit tests is **`test`**.
+
+==== External REST Tests 
+
+The `'elasticsearch.testclusters'` Gradle plugin has support for configuring
+and starting, and stopping clusters used for testing. 
+These test run against a cluster started from the generated distribution with 
+a configuration specific for the test and only interact using the REST APIs. 
+
+These can use the full default or oss distribution but we can also use a special
+one (integ-test) built for testing only that doesn't have any modules added.
+Module that are needed by the test can be added trough the testclusters plugin.
+This approach is preferred as it leads to better isolation, faster execution,
+and better build avoidance: if a change only changes a particular module, 
+all the others don't have to be re-tested, except when testing using the full 
+distribution.
+
+**`ESRestTestCase`**: is the base class for these tests.
+These tests can also be implemented using YAML and ran using a subclass of
+**ESClientYamlSuiteTestCase**.
+
+The Gradle target to run rest tests is **`integTest`**
+
+==== Internal Cluster Tests
+
+There are multiple base classes for these:
+
+* **`ESSingleNodeTestCase`**: This test case sets up a cluster that has a
+  single node.
+* **`ESIntegTestCase`**: An integration test case that creates a cluster that
+  might have multiple nodes.
+
+The reason why `ESSingleNodeTestCase` exists is that all our components used to
+be very hard to set up in isolation, which had led us to having a number of
+integration tests but close to no unit tests. `ESSingleNodeTestCase` is a
+workaround for this issue which provides an easy way to spin up a node and get
+access to components that are hard to instantiate like `IndicesService`.
+Whenever practical, you should prefer unit tests.
+
+Many tests extend `ESIntegTestCase`, mostly because this is how most tests used
+to work in the early days of Elasticsearch. However the complexity of these
+tests tends to make them hard to debug. Whenever the functionality that is
+being tested isn't intimately dependent on how Elasticsearch behaves as a
+cluster, it is recommended to write unit tests or REST tests instead.
+
+In short, most new functionality should come with unit tests, and optionally
+REST tests to test integration.
+
+==== Backwards Compatability Tests (BWC)
+
+Each version of Elasticsearch knows about the versions it's compatible with. 
+These tests run trough various upgrade scenarios from each compatibible version 
+to the current one. 
+This also includes snapshot versions (being worked on but not released).
+The build knows how to check out and build them so the tip of their branch
+can be tested.
+
+The Gradle target to run these is **`bwcTest`**
+
+==== OS / Packaging tests
+
+Thes live in **`qa/os`** and their goal is to verify that all the distribution
+types (archive, deb, rpm, windows installer) are compatible with the Operating
```

Comment:
We don't test the windows installer; that is managed completely externally from the elasticsearch project.